File: Recognized Patterns/Merge Sort/AJs/merge_sort.py
# two functions:
# merge sorted arrays
# recursively divide list into two and merge

def merge(list1, list2):
    # create a counter that loops over list1
    # compare that index between list1 and list2
    results = []
    i = 0
    j = 0
    while (i < len(list1)) & (j < len(list2)):
        if(list1[i] < list2[j]):
            results.append(list1[i])
            i+=1
        elif(list2[j] <= list1[i]):
            results.append(list2[j])
            j+=1
    
    while i < len(list1):
        results.append(list1[i])
        i+=1
    
    while j < len(list2):
        results.append(list2[j])
        j+=1
    
    return results
    
def mergeSort(a_list):
    if len(a_list) <= 1: return a_list

    half = len(a_list)//2
    
    left = mergeSort(a_list[:half])
    right = mergeSort(a_list[half:])

    return merge(left, right)
    


# merge([1,2,3],[3,6,7])
print(mergeSort([1,4,3,5,32,6,7,6,8,7,66,5,4,3,4,6]))
print(mergeSort([101,2,3,5,4,3,5,100,3,2,33,23,1,2,5,66,4,5,4,3,6,7,8,76,5,99,8,-6,0]))

File: Practice Library/Any Language/Sorting and Searching/Merge Sort/AJs/merge_sort.cs
using System.Collections;
using System.Double;

public List<int> Merge(List<int> a, List<int> b) {
    // merges two sorted array
    var results = new List<int>();
    int i=0;
    int j=0;
    
    while(i<a.Count() && j<b.Count()) {
        if(a[i] > b[j]) {
            results.Add(b[j]);
            j++;
        }
        else {
            results.Add(a[i]);
            i++;
        }
    }
    while(i<a.Count()) {
        results.Add(a[i]);
        i++;
    }

    while(j<b.Count()) {
        results.Add(b[j]);
        j++;
    }
    return results;
}

public List<int> MergeSort(List<int> nums) {
    if(nums.Count()<=1) return nums;
    
    int mid = nums.Count()/2;
    var a1 = new List<int>();
    var a2 = new List<int>();
    for (int i = 0; i < mid; i++)
    {
        a1.Add(nums[i]);
    }
    for (int i = mid; i < nums.Count(); i++)
    {
        a2.Add(nums[i]);
    }
    var left = MergeSort(a1);
    var right = MergeSort(a2);

    var ans = Merge(left, right);
    return ans;
}


Console.WriteLine(string.Join(",", MergeSort(new List<int> {1,3,4,2,4,5,6,4,1,33,6,78,8,9,5,0,-6})));
// Merge(new List<int> {1}, new List<int> {2});